### ISSUE TITLE
Transfrom CRLF to \n

### ISSUE THREAD

**Issue by halatmit** [mit-cml/appinventor-sources#208]
Change blocklyeditor/src/generators/yail.js, quotifyForREPL  so that CR followed by LF is treated the same as LF, i.e., transformed to an unescaped \n

This is to handle lines read in from Windows systems.
See forum issue

https://groups.google.com/forum/#!category-topic/mitappinventortest/tzoNdh4HyYY

**Comment by cdalsass** [comment]
I think the proper place to fix this is within the CSV parsing functions such as CsvUtil.java. quotifyForREPL deals with escaping strings from Blockly passed to REPL, and doesn't deal with what's inside CSV files. 

FYI: feel free to assign this issue to me.

**Comment by halatmit** [comment]
I agree:  It seems like a simple change to quotifyForREPL

**Comment by halatmit** [comment]
Assigned (and invited as a developer)



== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by cdalsass** [comment]
I don't think this is a bug. Rather, the original poster seems to have been confused about the API. (He should not have been putting CRs or LFs into a CSV "row" unless it was surrounded by quotes). This led to discussion which got somewhat off track from original problem. See my response at https://groups.google.com/d/msg/mitappinventortest/tzoNdh4HyYY/Q1hc3xMoUOgJ

**Comment by halatmit** [comment]
Whether or not it's the original issue with CSV, the problem to fix here arises from reading a file from a Windows machine, which ends lines with CR LF, unlike Mac and Linux, which ends them with just LF.   The solution we decided on is for the parser to transform occurrences of CR followed by LF into and unescaped LF, just as if it had read an LF without the CR.

**Comment by cdalsass** [comment]
I do understand the problem. However, I disagree that quotifyForREPL in yail.js is the best place to make changes. (But maybe I am misunderstanding the code - I apologize if that's the case). 

Question: Has there has been a persistent problem in AI2 with newline issues, prior to this post? (From reading the original Google post, it appears to me the bug was just noticed in December - which contradicts that.)

In my test case, I show that both Windows and Unix files are both handled properly in the context of CSV row parsing. But I think you're saying that **any time a file is read**, we should normalize return characters. 

If you want that, the GotText event is probably the correct place to make changes. That event is raised during reads from a web resource (Web.java) or a regular file (File.java) or anything else, and I'm concerned about inadvertent bugs and separation of concerns. Perhaps we apply the transformation when that event is raised, using a separate TextTransformation class. 

Also, is there a test case for this bug? I can use Taifun's case https://groups.google.com/d/msg/mitappinventortest/tzoNdh4HyYY/iqu73m8MVskJ , but am not sure his case occurs often enough to justify across-the-board changes.

Thanks for your time and patience as I work through this issue.

**Comment by halatmit** [comment]
Taifun, Lyn:

What do you think about Charlie's comment?



== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by puravidaapps** [comment]
I don't know, if the problem already existed earlier, but this actually is not relevant, is it?
I'm sorry, I can't comment on where is the best place to make changes in the code. 
Btw. in the forum there is currently another discussion about \n, see here https://groups.google.com/d/msg/mitappinventortest/bwLwrGNk-Mo/PnjPJNa9PIwJ which probably also  is relevant for this issue? 
Taifun

**Comment by halatmit** [comment]
+ charlie, lyn




== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by cdalsass** [comment]
Thanks so much for the feedback, Taifun!

If there are not ongoing problems reported with Window and Mac compatibility, it may be better use of time to focus on other bugs. I don't see a trivial, yet clean (e.g. well-designed) solution to this problem. Still open to other suggestions. (Might be better to discuss in person).

p.s. the link you provided appears to be a different issue.

**Comment by halatmit** [comment]
How did you decide to handle this?




== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by cdalsass** [comment]
Hal, if you are asking me (sorry, but I can't tell who your question is directed to)...  'close as not a bug' unless there are persistent problems with Windows/Unix text files. If there are ongoing problems, I'd recommend you reopen as new bug describing the problem more generally. I'd also add some simple test cases, so we can be sure it's fixed.

**Comment by halatmit** [comment]
Taifun,

Would you like
to comment?

**Comment by halatmit** [comment]
Taifun,

Could you comment on this?  My understanding is that we need this change in
App Inventor in order for people to read files, independently of whether
the files were produced by Mac/Linux vs. Windows.

Is that your understanding?

Charlie, if you don't think this fix is necessary, how would you propose to
deal with the issue?




== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by puravidaapps** [comment]
yes, exactly.

From a user point of view it doesn't matter if a file was created with Windows or Linux/Mac, the user only wants to get rid of the invisible newline character (be it CRLF or LF only)
so this currently works only for Linux/Mac files (LF), but not for Windows files (CRLF)...
If I use the replace all block to replace \n, I would expect to get the newline character(s) removed (be it CRLF or LF only)

![unbenannt](https://cloud.githubusercontent.com/assets/2848474/5792477/760127a2-9edf-11e4-9d80-76770f034f4f.PNG)

you can find the test project here https://groups.google.com/d/msg/mitappinventortest/tzoNdh4HyYY/iqu73m8MVskJ

btw. see also the operating system market share here http://www.netmarketshare.com/operating-system-market-share.aspx?qprid=10&qpcustomd=0
according to this report, Windows has a market share of ~90%...

comment by ABG: Postel's Law https://www.google.com/search?q=Postel%27s+Law&oq=Postel%27s+Law would advise accepting both LF and CRLF but emitting just LF. Since Android is a member of the unix family, the common convention on that platform would be to emit LF.

Taifun

**Comment by halatmit** [comment]
I find Taifun's comment convincing.   Charlie, do you want to take this on, or should someone else do it?

**Comment by cdalsass** [comment]
OK. 

I'd like to finish it up. 

I'll probably modify ReadFrom -> AsyncRead() in File.java line line 261. Trivial to replace text directly there as I have it in memory as String. 

I wonder if we should include a backward-compatibility option for case (a) below. Do you think that's an issue knowing what scripts people have developed?

I am still have some concerns:

a) I bet will there will be a few users who are activity detecting for Windows files. Their scripts break with this implementation.
b) someone who is technical enough to understand \n is probably technical enough to understand \r. (We want to make things easy, but should we "unteach" people by saying \n is same as \r\n?)
c) CSV parsing routines already take different file types into account and do work now. 
d) We won't be normalizing new lines for files read from say, a URL.

**Comment by halatmit** [comment]
I was imagining putting the change in quotifyForREPL because that's the
same place we worry about backslashed \n, and it seems safer to localize
all of that transformation stuff in the same place.

I don't know what the implications are of putting it in file.java.   What
happens, for example, if a user types

abc\r\ndef

into a text box, retrieves the contents, and hits DoIt.  What is shown in
the balloon?








== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by cdalsass** [comment]
Yes, I am also worried about unexpected bugs (if I modify File.java). 

Maybe you are right all along, saying quotifyForREPL is the best place to modify. 

This solution doesn't feel right to me because it means making changes in a place completely unrelated to text files, violating single responsibility principle and confusing developers in the future. "I'm quotifying for REPL, but I also just happen to be solving the Windows/Mac newline problem"?  

Here's a different idea - what about normalizing newlines when a (text) file is uploaded via web interface?

I'll review how file contents are passed through to the REPL, and reconsider. Thanks for the patience. 

If I have more questions, I'll connect offline - don't want to go in circles here.

**Comment by halatmit** [comment]
Thanks, take a look.

I'd feel safer if all the backslashy stuff was handled at the same place in
the code.

But let me know.


== Hal Abelson
hal@mit.edu
Prof. of Comp. Sci. and Eng.
MIT Dept. of Elec. Eng. and Comp. Sci.
MIT Media Lab, Second Class Faculty

**Comment by cdalsass** [comment]
I took another look at quotifyForREPL this morning. 

I alert()'d the arguments passed to quotifyForREPL and exercised the functionality of uploading a media file, reading it etc. At no time are the **contents of media files** ever passed into this function. (Perhaps this point is where there is confusion in our discussions. Perhaps you think the file contents are entered in a text box? They are not in the test case I have.*). Only the user (developer/script author) entered strings are passed into this function. So I am very sure now that quotifyForREPL is not the place to modify the function. 

*Just to clarify. In the example M.aia provided by Taifun, a Windows text file was uploaded into to the "Media" section. Next blocks for ReadFrom and GotText() calls were used to read contents of the file.

I also researched how media files are transferred from web interface to sdcard using Companion. In replmgr.js, line 1000, the putAsset function generates PUT requests for all media files user has uploaded. AI Companion captures these put requests and stores the files on the sd card, synching the assets. I noticed that the PUT requests are generated after a change to the blocks. 

So I still think the best place to make this change is in File.java, 261 or line 294. A one-liner there will solve it. 

s = s.replaceAll("\r\n", "\n'");

This will only affect files that are opened locally (vs. URLS etc). Contents entered into text boxes will not be affected. I am still concerned about bugs with existing scripts, however, such as people that actually want to detect Windows files, and don't expect \r\n to go away.
